quinn-proto/src/endpoint.rs

```diff
@@ -83,6 +83,9 @@ pub struct Config {
     pub loss_reduction_factor: u16,
 
     pub tls_server_config: Arc<ServerConfig>,
+
+    /// Length of connection IDs for the endpoint
+    pub local_cid_len: usize,
```

Comment:
I'm not so much interested in the CPU performance overhead, but I was thinking the bandwidth overhead might be nice to elide where possible (for example, in many clients). I've added a comment to that effect and the assertion as you suggested.

I think avoiding the maintenance of `self.connection_ids` will clutter up the code and may not pay for itself in performance with any added branches, so I'd prefer to keep it. However, we can check this as a special case in the `new_cid()` method so that it can raise an error.

Reply:
The bandwidth overhead is what I'm referring to, yeah. The important thing is that we don't try to look up 0-length CIDs in `handle_decode`.

Reply:
I don't think the lookup is a problem as long as we restrict the endpoint to only ever giving out the zero-length CID once, which I think my updated patch covers. Take a look.

Reply:
Doesn't that restrict the endpoint to having at most one concurrent connection? §5.1 suggests that we should use the remote address+port to identify connections when zero-length CIDs are in use.

Reply:
Yup, you're right. I'll tweak this more.

Reply:
Should be better now. Added this:

```diff
diff --git a/quinn-proto/src/endpoint.rs b/quinn-proto/src/endpoint.rs
index bee4054..d85842a 100644
--- a/quinn-proto/src/endpoint.rs
+++ b/quinn-proto/src/endpoint.rs
@@ -333,12 +333,16 @@ impl Endpoint {
         //

         let dst_cid = partial_decode.dst_cid();
-        let conn = self
-            .connection_ids
-            .get(&dst_cid)
-            .or_else(|| self.connection_ids_initial.get(&dst_cid))
-            .or_else(|| self.connection_remotes.get(&remote))
-            .cloned();
+        let conn = {
+            let conn = if self.ctx.config.local_cid_len > 0 {
+                self.connection_ids.get(&dst_cid)
+            } else {
+                None
+            };
+            conn.or_else(|| self.connection_ids_initial.get(&dst_cid))
+                .or_else(|| self.connection_remotes.get(&remote))
+                .cloned()
+        };
         if let Some(conn) = conn {
             return self.connections[conn.0].handle_decode(
                 &mut self.ctx,
@@ -489,7 +493,9 @@ impl Endpoint {
             ));
             conn
         };
-        self.connection_ids.insert(local_id, conn);
+        if self.ctx.config.local_cid_len > 0 {
+            self.connection_ids.insert(local_id, conn);
+        }
         self.connection_remotes.insert(remote, conn);
         conn
     }
@@ -617,8 +623,10 @@ impl Endpoint {
             self.connection_ids_initial
                 .remove(&self.connections[conn.0].init_cid);
         }
-        self.connection_ids
-            .remove(&self.connections[conn.0].loc_cid);
+        if self.ctx.config.local_cid_len > 0 {
+            self.connection_ids
+                .remove(&self.connections[conn.0].loc_cid);
+        }
         self.connection_remotes
             .remove(&self.connections[conn.0].remote);
         self.ctx.dirty_conns.remove(&conn);
```